contracts/guilds/lib/math_utils.cairo

```diff
@@ -0,0 +1,81 @@
+%lang starknet
+
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.uint256 import (
+    Uint256,
+    uint256_add,
+    uint256_sub,
+    uint256_lt,
+    uint256_eq,
+    uint256_mul
+)
+
+@view
+func array_sum{
+        syscall_ptr : felt*, 
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr, 
+    }(
+        arr_len: felt,
+        arr: felt*
+    ) -> (sum: felt):
+    if arr_len == 0:
+        return (sum=0)
+    end
+
+    let (sum_of_rest) = array_sum(arr_len=arr_len - 1, arr=arr + 1)
```

Comment:
I don't think this will be tail call optimized, because there's one more operations (`[]` and `+`) after the recursive call. I might be wrong of course. Do you know if the compiler is smart enough to deal with this?

Reply:
I took this from [hello_cairo](https://www.cairo-lang.org/docs/hello_cairo/intro.html). I think it is ok, they provide more description on why they implement this way.